torch/csrc/distributed/c10d/reducer.cpp

```diff
@@ -1079,6 +1124,23 @@ std::vector<std::vector<size_t>> Reducer::rebuildBuckets() {
   return rebuilt_bucket_indices;
 }
 
+// See Note [DDP Communication Hook]
+void Reducer::register_comm_hook(py::object state, py::object comm_hook) {
+  Reducer::register_comm_hook_internal(
+      std::make_unique<PythonCommHook>(std::move(state), std::move(comm_hook)));
+}
+
```

Comment:
Right, we should avoid python dependency in reducer.cpp. I think calling `register_comm_hook_internal` from `comm.h` may not possible because this may circular dependency between `comm.h` and `reducer.h`. Instead, I created a static function in `init.cpp` and call it from DDP Python API as `dist.PythonHookBinder.register_comm_hook(self.reducer, state, hook)`.